lib/apiservers/engine/backends/convert/state.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/go-units"
+
+	"github.com/vmware/vic/lib/apiservers/portlayer/models"
+	"github.com/vmware/vic/pkg/trace"
+)
+
+// State will create and return a docker ContainerState object
+// from the passed vic ContainerInfo object
+func State(info *models.ContainerInfo) *types.ContainerState {
+	defer trace.End(trace.Begin(""))
+
+	// ensure we have the data we need
+	if info == nil || info.ProcessConfig == nil || info.ContainerConfig == nil {
+		return nil
+	}
+
+	dockerState := &types.ContainerState{}
+
+	// convert start / stop times
+	var started time.Time
+	var finished time.Time
+
+	if info.ProcessConfig.StartTime > 0 {
+		started = time.Unix(info.ProcessConfig.StartTime, 0)
+		dockerState.StartedAt = time.Unix(info.ProcessConfig.StartTime, 0).Format(time.RFC3339Nano)
+	}
+
+	if info.ProcessConfig.StopTime > 0 {
+		finished = time.Unix(info.ProcessConfig.StopTime, 0)
+		dockerState.FinishedAt = time.Unix(info.ProcessConfig.StopTime, 0).Format(time.RFC3339Nano)
+	}
+
+	// set docker status to state and we'll change if needed
+	dockStatus := info.ContainerConfig.State
+	// set exitCoe and change if needed
+	exitCode := int(info.ProcessConfig.ExitCode)
+
+	switch info.ContainerConfig.State {
+	case "Running":
+		// if we don't have a start date leave the status as the state
+		if !started.IsZero() {
+			dockStatus = fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(started)))
+			dockerState.Running = true
+		}
+	case "Stopped":
+		// if we don't have a finished date then don't process exitCode and return "Stopped" for the status
+		if !finished.IsZero() {
```

Comment:
If the cVM is powered off out-of-band (host outage or otherwise) is the state correctly set to stopped?
Is it possible to use a message that informs the user that exit status and time are unavailable rather than just omit the details? Any idea what docker says if the containerd handler is SIGKILLed?